src/nwp_consumer/internal/entities/tensorstore.py

```diff
@@ -0,0 +1,501 @@
+        except Exception as e:
+            return Failure(OSError(
+                "Cannot determine missing times in store due to "
+                f"error reading '{self.path}': {e}",
+            ))
+        missing_times: list[dt.datetime] = []
+        for it in store_da.coords["init_time"].values:
+            if store_da.sel(init_time=it).isel({
+                d: slice(0, 2) for d in self.coordinate_map.dims
+                if d != "init_time"
+            }).isnull().all().values:
+                missing_times.append(pd.Timestamp(it).to_pydatetime().replace(tzinfo=dt.UTC))
+        return Success(missing_times)
+
+    @staticmethod
+    def _create_zarrstore_s3(s3_folder: str, filename: str) \
+            -> ResultE[tuple[MutableMapping, str]]:  # type: ignore
+        """Create a mutable mapping to an S3 store.
+
+        Authentication with S3 is done via botocore's credential discovery.
+
+        Returns:
+            A tuple containing the store mapping and the path to the store,
+            in a result object indicating success or failure.
+
+        See Also:
+          - https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials
+        """
+        import s3fs
+        if not s3_folder.startswith("s3://"):
+            return Failure(ValueError(
+                "S3 folder path must start with 's3://'. "
+                f"Got: {s3_folder}",
+            ))
+        log.debug("Attempting AWS connection using credential discovery")
+        try:
+            fs = s3fs.S3FileSystem(
+                anon=False,
+                client_kwargs={
+                    "region_name": os.getenv("AWS_REGION", "eu-west-1"),
+                    "endpoint_url": os.getenv("AWS_ENDPOINT_URL", None),
+                },
+            )
+            path = s3_folder + "/" + filename
+            fs.mkdirs(path=path, exist_ok=True)
+            store = s3fs.mapping.S3Map(path, fs, check=False, create=True)
+        except Exception as e:
+            return Failure(OSError(
+                f"Unable to create file mapping for path '{path}'. "
+                "Ensure ZARRDIR environment variable is specified correctly, "
+                "and AWS credentials are discoverable by botocore. "
+                f"Error context: {e}",
+            ))
+        return Success((store, path))
+
+    @staticmethod
+    def gen_store_filename(coords: NWPDimensionCoordinateMap) -> str:
+        """Create a filename for the store.
+
+        If the store only covers a single init_time, the filename is the init time.
+        Else, if it covers multiple init_times, the filename is the range of init times.
+        The extension is '.zarr'.
+        """
+        store_range: str = coords.init_time[0].strftime("%Y%m%d%H")
+        if len(coords.init_time) > 1:
+            store_range = f"{coords.init_time[0]:%Y%m%d%H}-{coords.init_time[-1]:%Y%m%d%H}"
```

Comment:
could you do `store_range+=f"-{coords.init_time[-1]:%Y%m%d%H}"` to shortened the code